10_rest/willshen8/cmd/puppy-server/main.go

```diff
@@ -0,0 +1,118 @@
+package main
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"log"
+	"net/http"
+	"os"
+	"strconv"
+
+	"github.com/anz-bank/go-course/10_rest/willshen8/pkg/puppy"
+	"github.com/go-chi/chi"
+	"github.com/go-chi/chi/middleware"
+	"gopkg.in/alecthomas/kingpin.v2"
+)
+
+var out io.Writer = os.Stdout
+
+var (
+	args     = os.Args[1:]
+	fileName = kingpin.Flag("data", "Puppy Json File Name").Short('d').ExistingFile()
+	port     = kingpin.Flag("port", "Port of the server").Short('p').Default("8888").Int()
+	storer   = kingpin.Flag("store", "Map/Sync Store").Short('s').Default("map").String()
+)
+
+// Config contains arguments parsed from commandline flags.
+type Config struct {
+	file   string
+	port   int
+	storer string
+}
+
+func main() {
+	cmdConfigs := parseCmdArgs(args)
+	newStorer, _ := createStorer(cmdConfigs)
+	decodedPuppies, _ := parseFileFlag(cmdConfigs)
+	portFlag, portErr := parsePortFlag(cmdConfigs)
+	if portErr != nil {
+		fmt.Println(portErr)
+	}
+	port := ":" + portFlag
+
+	puppyStorer, _ := createPuppies(newStorer, decodedPuppies)
+	fmt.Fprintln(out, puppyStorer)
+	puppyHandler := puppy.NewRestHandler(puppyStorer)
+	router := SetupRouter()
+	puppy.SetupRoutes(router, *puppyHandler)
+	log.Fatal(http.ListenAndServe(port, router))
+}
+
+// creatStorer will take the commandline arg and create the appropriate store
+func createStorer(c *Config) (puppy.Storer, error) {
+	switch c.storer {
+	case "map":
+		return puppy.NewMapStore(), nil
+	case "sync":
+		return puppy.NewSyncStore(), nil
+	}
+	return nil, errors.New("map/sync are the only acceptable flag values")
+}
+
+func parseCmdArgs(args []string) *Config {
+	if _, parseError := kingpin.CommandLine.Parse(args); parseError != nil {
+		panic(parseError)
+	}
+	return &Config{file: *fileName, port: *port, storer: *storer}
+}
+
+func parseFileFlag(c *Config) ([]puppy.Puppy, error) {
+	var file *string
+	var puppies []puppy.Puppy
+	file = &c.file
+	jsonData := readFile(*file)
+
+	if err := json.Unmarshal(jsonData, &puppies); err != nil {
+		return nil, err
+	}
+	return puppies, nil
+}
+
+//parsePortFlag check for valid port number entered from commandLine
```

Comment:
in some places you've got space after `//` in some not. `grep -r "^//"`.
This is a nit, feel free to ignore this if you want.

Reply:
Good point, I like consistency too. Not sure how that happened.